Default Ingress Passthrough for KubeVirt provider

davidvossel opened the pull request:
# Overview

This PR adds the new `hostedCluster.Spec.Platform.KubeVirt.BaseDomainPassthrough`  option, which automatically sets the HostedCluster's BaseDomain to a unique subdomain of the mgmt/infra cluster's `*apps` dns. The HCCO operator in turn sets up the correct services and routes to pass through traffic on that subdomain to the KubeVirt VMs.

We've already been using this technique for months now both in our manual development of the KubeVirt Platform and CI [[1]](https://github.com/openshift/release/blob/1b66770683fa462c3cdc0e1dacbe6796215edf9b/ci-operator/step-registry/hypershift/kubevirt/create/hypershift-kubevirt-create-chain.yaml#L58). The technique is also documented in our hypershift upstream docs here [[2]](https://hypershift-docs.netlify.app/how-to/kubevirt/create-kubevirt-cluster/ ). We're even seeing third parties document manually setting up this same technique[ [3]](https://almogelfassy.medium.com/cluster-as-a-service-the-next-phase-of-development-for-kubernetes-by-hypershift-4f1b8d28ad78 ).

What we know now is that this ingress passthrough technique is the primary way users evaluate the Hypershift KubeVirt platform. It's the easiest way to get a functioning cluster. The `hostedCluster.Spec.Platform.KubeVirt.BaseDomainPassthrough` option improves the user experience for people who would otherwise be manually deploying the same logic.

# Design

The figure below outlines how this technique works. **The orange boxes are the route and service that the HCCO manages in the infra/mgmt cluster which passes through traffic to the kubevirt guest cluster.**

![kubevirt hypershift default route ingress notes](https://user-images.githubusercontent.com/683713/194150896-557ce5d8-1e62-402c-9911-705700cdfb72.jpg)

# Usage

With this feature, users of the `hypershift` cli simply omit the basedomain when creating their kubevirt platform cluster and that automatically renders a HostedCluster with the passthrough feature enabled. Below is an example of what the cli looks like that uses the automatic passthrough default ingress.

```
bin/hypershift create cluster kubevirt \
	--name $CLUSTER_NAME \
	--pull-secret $PULL_SECRET_PATH \
	--ssh-key $SSH_KEY_PATH  \
```

# References

1. https://github.com/openshift/release/blob/1b66770683fa462c3cdc0e1dacbe6796215edf9b/ci-operator/step-registry/hypershift/kubevirt/create/hypershift-kubevirt-create-chain.yaml#L58
2. https://hypershift-docs.netlify.app/how-to/kubevirt/create-kubevirt-cluster/ 
3. https://almogelfassy.medium.com/cluster-as-a-service-the-next-phase-of-development-for-kubernetes-by-hypershift-4f1b8d28ad78

davidvossel commented:
@orenc1 maybe you could give this a review? We'll get one of the hypershift approvers to take a look once we're happy with it

davidvossel commented:
/retest

orenc1 commented on cmd/cluster/kubevirt/create.go:
```diff
@@ -102,5 +96,12 @@ func applyPlatformSpecificsValues(ctx context.Context, exampleOptions *apifixtur
+	if opts.BaseDomain != "" {
+		exampleOptions.BaseDomain = opts.BaseDomain
+	} else {
+		exampleOptions.Kubevirt.BaseDomainPassthrough = true
```
don't we want the default for `BaseDomainPassthrough` boolean property be set in the CRD level?

davidvossel replied:
I'm not sure if it's possible to default the value accurately.

The issue is that when BaseDomainPassthrough is set, the hc.Spec.DNS.BaseDomain must be empty so the controller can automatically set it. So, if we default BaseDomainPassthrough=true on the CRD, and the user also set's the DNS.BaseDomain, then the configuration is invalid (unless they set the base domain perfectly to what the passthrough logic would have set it too).

So, we can only default BaseDomanPassthrough when there is no BaseDomain set by the user. I can default this in the CLI, but i don't think i can express that complex default behavior on the CRD today... at least I haven't seen something like that done yet. I've heard the community is improving what can be expressed on the CRD, but i haven't seen it yet.

orenc1 commented on api/v1alpha1/hostedcluster_types.go:
```diff
@@ -650,6 +650,40 @@ type PlatformSpec struct {
+	//   Infra/Mgmt cluster's DNS
+	//     Base: example.com
+	//     Cluster: mgmt-cluster.example.com
+	//     Apps:    *apps.mgmt-cluster.example.com
```
since the wildcard subdomain is under `apps` exactly, i recommend changing:
`*apps` to `*.apps` in all occurrences to make it more evident the url `*.something-apps.mgmt-cluster.example.com` it not valid.

davidvossel replied:
fixed

orenc1 commented on control-plane-operator/controllers/hostedcontrolplane/configoperator/reconcile.go:
```diff
@@ -114,9 +114,23 @@ func ReconcileRole(role *rbacv1.Role, ownerRef config.OwnerRef) error {
 				"secrets",
+			},
+			Verbs: []string{
+				"create",
```
why did you add "create secrets" permission?
I don't see any new secrets creation in the new code.

davidvossel replied:
ah, good catch. I don't think we need `create`.

orenc1 reviewed (commented):
Good job and nicely done @davidvossel ,
This will pretty much improve user experience and offload the need to create a service and a route to access the kubevirt guest cluster.

I've put some minor comments/questions inline.
And I have one more question/thought:
It is technically possible to configure a custom IngressController in a cluster to forward some of the applications/workloads in another subdomain, or to use it as an internal load balancer.
in the current implementation, the base domain will be hardcoded to `*.apps.mgmt-cluster...`
Do we want to allow usage of custom ingress controllers for the guest clusters in the future?

davidvossel commented on control-plane-operator/hostedclusterconfigoperator/controllers/resources/ingress/reconcile.go:
```diff
@@ -92,3 +95,58 @@ func ReconcileDefaultIngressControllerCertSecret(certSecret *corev1.Secret, sour
+	}
+
+	service.Spec = corev1.ServiceSpec{
+		Ports: []corev1.ServicePort{
```
I don't have a good explanation for this yet, but adding port 80 and 443 at the same time to the service breaks the console operators health check. This might be something we need to investigate independently.

Because of this weirdness, I propose that we use an annotation to toggle on the passthrough rather than adding it to the api right now. That way we can back out this functionality later on without breaking API compatiblity if it turns out we can't support 443 and 80 at the same time. I'll make this change.


I did investigate this briefly, and i'm a bit stumped. The console url works fine from my browser but the console health check returns the following error.
`
console   4.12.0-ec.5   False       False         False      13m     RouteHealthAvailable: failed to GET route (https://console-openshift-console.apps.dvossel-hyper1.apps.dvossel-test58.cnv.azure.devcluster.openshift.com): Get "https://console-openshift-console.apps.dvossel-hyper1.apps.dvossel-test58.cnv.azure.devcluster.openshift.com": http: server gave HTTP response to HTTPS client
`

I even made some golang code to test the url and it returns that TLS is enabled.

```
package main

import (
        "crypto/tls"
        "fmt"
        "net/http"
)

func main() {
        client := &http.Client{
                Transport: &http.Transport{
                        TLSClientConfig: &tls.Config{
                                InsecureSkipVerify: true,
                        },
                },
        }
        req, _ := http.NewRequest(http.MethodGet, "https://console-openshift-console.apps.dvossel-hyper1.apps.dvossel-test58.cnv.azure.devcluster.openshift.com:443", nil)
        resp, err := client.Do(req)
        if err != nil {
                panic(err)
        }
        fmt.Printf("resp: %s\n", *resp)
        if resp.TLS != nil {
                fmt.Printf("GOT TLS RESPONSE\n")
        }
}
```

davidvossel replied:
Wow, i tested this some more and figured out something even more strange.

If i curl the https url from within the guest cluster, it actually does report that http response to an https request... However, outside of the guest cluster it works fine.

davidvossel commented:
/hold

I'm going to make this option an annotation for now instead of adding it directly to the HostedCluster api. We'd like to have this functionality for people to evaluate HyperShift easily, but we're not quite yet ready to say this is something we're 100% confident we can support long term. Because of that uncertainty, an annotation would allow us to back the feature out without impacting the HostedCluster's API compatiblity.

davidvossel commented:
@orenc1 if you can, take another pass at this.

I changed this logic to be triggered by an annotation instead of adding a field to the HostedCluster API. That lets us begin using this logic with the hypershift cli now to ease people's evaluation of hypershift/kubevirt, while giving us an out if we find this is not supportable long term for some reason (or if we can't figure out how to get routing for port 443 and 80 at the same time) For now i left the PR with only exposing port 443 since exposing 443 and 80 seemed to cause strange behavior. We'll have to look more into that at some point.

I'm leaving the hold because I want to get someone from the hypershift team to confirm my annotation technique and RBAC usage is acceptable

orenc1 commented:
> @orenc1 if you can, take another pass at this.
> 
> I changed this logic to be triggered by an annotation instead of adding a field to the HostedCluster API. That lets us begin using this logic with the hypershift cli now to ease people's evaluation of hypershift/kubevirt, while giving us an out if we find this is not supportable long term for some reason (or if we can't figure out how to get routing for port 443 and 80 at the same time) For now i left the PR with only exposing port 443 since exposing 443 and 80 seemed to cause strange behavior. We'll have to look more into that at some point.
> 
> I'm leaving the hold because I want to get someone from the hypershift team to confirm my annotation technique and RBAC usage is acceptable

Reviewed, lgtm.
I'm fine with setting the annotation on the HostedControlPlane CR to indicate auto-creation of ingress service and route.

btw, why I'm not seeing the new passthrough [service](https://gcsweb-ci.apps.ci.l2s4.p1.openshiftapps.com/gcs/origin-ci-test/pr-logs/pull/openshift_hypershift/1787/pull-ci-openshift-hypershift-main-kubevirt-e2e-kubevirt-gcp-ovn/1588220830851010560/artifacts/e2e-kubevirt-gcp-ovn/kv-e2e/artifacts/TestCreateCluster_PreTeardownClusterDump/namespaces/e2e-clusters-9x99k-example-7zgqg/core/services/) and [route](https://gcsweb-ci.apps.ci.l2s4.p1.openshiftapps.com/gcs/origin-ci-test/pr-logs/pull/openshift_hypershift/1787/pull-ci-openshift-hypershift-main-kubevirt-e2e-kubevirt-gcp-ovn/1588220830851010560/artifacts/e2e-kubevirt-gcp-ovn/kv-e2e/artifacts/TestCreateCluster_PreTeardownClusterDump/namespaces/e2e-clusters-9x99k-example-7zgqg/route.openshift.io/routes/) in the job artifacts that should be created by hypershift operator?

orenc1 commented:
/retest

davidvossel commented:
/test kubevirt-e2e-kubevirt-aws-ovn

davidvossel commented:
/test kubevirt-e2e-kubevirt-aws-ovn

davidvossel commented on cmd/cluster/kubevirt/create.go:
```diff
@@ -104,5 +99,15 @@ func applyPlatformSpecificsValues(ctx context.Context, exampleOptions *apifixtur
+		if exampleOptions.Annotations == nil {
+			exampleOptions.Annotations = map[string]string{}
+		}
+		exampleOptions.Annotations[hyperv1.KubeVirtBaseDomainPassthroughAnnotation] = "true"
```
We're using an annotation here to signal to the controllers that automatic basedomain passthrough is in use. I picked an annotation here instead of using a new API field for this signal because I'm not yet ready to officially say "how" we'll enable this feature.

enxebre replied:
I'm not sure that annotation effectively give us anything else than opaqueness https://github.com/openshift/hypershift/pull/1787/files#r1067315404. I'm +1 to field.

davidvossel commented on control-plane-operator/hostedclusterconfigoperator/controllers/resources/resources.go:
```diff
@@ -719,6 +718,38 @@ func (r *reconciler) reconcileIngressController(ctx context.Context, hcp *hyperv
 		}
 	}
+
+	// Default Ingress is passed through as a subdomain of the infra/mgmt cluster
+	// for KubeVirt when the base domain passthrough feature is in use.
+	baseDomainPassthrough := hcp.Annotations[hyperv1.KubeVirtBaseDomainPassthroughAnnotation]
+	if hcp.Spec.Platform.Type == hyperv1.KubevirtPlatform && baseDomainPassthrough == "true" {
+		defaultIngressNodePortService := manifests.IngressDefaultIngressNodePortService()
+		err := r.client.Get(ctx, client.ObjectKeyFromObject(defaultIngressNodePortService), defaultIngressNodePortService)
+		if err != nil {
+			errs = append(errs, fmt.Errorf("failed to retrieve guest cluster ingress NodePort: %w", err))
+		}
+
+		hcpNamespace := hcp.Namespace
+
+		// Manifests for infra/mgmt cluster passthrough service
+		cpService := manifests.IngressDefaultIngressPassthroughService(hcpNamespace)
+
+		// Manifests for infra/mgmt cluster passthrough routes
+		cpPassthroughRoute := manifests.IngressDefaultIngressPassthroughRoute(hcpNamespace)
+
+		if _, err := r.CreateOrUpdate(ctx, r.cpClient, cpService, func() error {
+			return ingress.ReconcileDefaultIngressPassthroughService(cpService, defaultIngressNodePortService, hcp)
+		}); err != nil {
+			errs = append(errs, fmt.Errorf("failed to reconcile kubevirt ingress passthrough service: %w", err))
+		}
+
+		if _, err := r.CreateOrUpdate(ctx, r.cpClient, cpPassthroughRoute, func() error {
+			return ingress.ReconcileDefaultIngressPassthroughRoute(cpPassthroughRoute, cpService, hcp)
+		}); err != nil {
+			errs = append(errs, fmt.Errorf("failed to reconcile kubevirt ingress passthrough route: %w", err))
+		}
+	}
+
```
I'm creating a route and service in the hosted control plane namespace while in the HCCO (which typically only works on the guest client, not the mgmt client).

This is being done because I have to get the NodePort's port used by the default ingress services in the guest cluster in order to create the service in the mgmt cluster. basically, the mgmt cluster service has to point the backend "somewhere", and that somewhere is the nodeport's port of the routers in the guest cluster.

davidvossel replied:
done

davidvossel commented on hypershift-operator/controllers/hostedcluster/hostedcluster_controller.go:
```diff
@@ -4212,6 +4218,48 @@ func (r *HostedClusterReconciler) serviceAccountSigningKeyBytes(ctx context.Cont
+			return err
+		}
+
+		// For the KubeVirt platform, the basedomain can be autogenerated using
+		// the *.apps domain of the management/infra hosting cluster
+		// This makes the guest cluster's base domain a subdomain of the
+		// hypershift infra/mgmt cluster's base domain.
+		//
+		// Example:
+		//   Infra/Mgmt cluster's DNS
+		//     Base: example.com
+		//     Cluster: mgmt-cluster.example.com
+		//     Apps:    *.apps.mgmt-cluster.example.com
+		//   KubeVirt Guest cluster's DNS
+		//     Base: apps.mgmt-cluster.example.com
+		//     Cluster: guest.apps.mgmt-cluster.example.com
+		//     Apps: *.apps.guest.apps.mgmt-cluster.example.com
+		//
+		// This is possible using OCP wildcard routes
+		hc.Spec.DNS.BaseDomain = defaultIngressOperator.Status.Domain
```
when no basedomain is set, we're detecting the mgmt cluster's default ingress domain and using that here.

davidvossel commented on hypershift-operator/controllers/hostedcluster/hostedcluster_webhook.go:
```diff
@@ -260,6 +260,17 @@ func validateHostedClusterUpdate(new *hyperv1.HostedCluster, old *hyperv1.Hosted
 		old.Spec.Networking.APIServer.Port = new.Spec.Networking.APIServer.Port
 	}
 
+	// We default the basedomain for KubeVirt clusters, so we allow the baseDomain
+	// to go from unset to set, but no updates after basedomain is set.
+	baseDomainPassthrough := new.Annotations[hyperv1.KubeVirtBaseDomainPassthroughAnnotation]
+	if new.Spec.Platform.Type == hyperv1.KubevirtPlatform &&
+		baseDomainPassthrough == "true" &&
+		new.Spec.DNS.BaseDomain != "" &&
+		old.Spec.DNS.BaseDomain == "" {
+
+		old.Spec.DNS.BaseDomain = new.Spec.DNS.BaseDomain
+	}
+
```
BaseDomain is typically immutable. I had to make an exception when the BaseDomain is empty and the kubevirt platform is in use. This is how we default the BaseDomain.

davidvossel commented on test/e2e/util/kubevirt.go:
```diff
@@ -81,80 +76,3 @@ func WaitForKubeVirtCluster(t *testing.T, ctx context.Context, client crclient.C
 
 	t.Logf("KubeVirtCluster is ready in %s", time.Since(start).Round(time.Second))
 }
-
-func createKubeVirtClusterWildcardRoute(t *testing.T, ctx context.Context, client crclient.Client, guestClient crclient.Client, hostedCluster *hyperv1.HostedCluster, baseDomain string) {
-
-	g := NewWithT(t)
-
```
we were doing all this DNS/route passthrough logic in the e2e tests. Now that can be stripped out. that's why all this test code is being removed.

davidvossel reviewed (commented):
I added some comments inline to help hypershift reviewers understand the key parts of this pr

enxebre commented on api/v1beta1/hostedcluster_types.go:
```diff
@@ -124,6 +124,27 @@ const (
+	//     Cluster: guest.apps.mgmt-cluster.example.com
+	//     Apps: *.apps.guest.apps.mgmt-cluster.example.com
+	//
+	// This is possible using OCP wildcard routes
```
PR desc mention an API field but the PR implements an annotation.
Annotations used as in this PR are just opaque APIs that still need to follow support policies once we ship. Because of this I'm +1 to make this a field.

enxebre commented on control-plane-operator/controllers/hostedcontrolplane/configoperator/reconcile.go:
```diff
@@ -151,6 +151,51 @@ func ReconcileRole(role *rbacv1.Role, ownerRef config.OwnerRef) error {
+		// By isolating these rules behind the KubevirtPlatform switch case,
+		// we know we can add/remove from this list in the future without
+		// impacting other platforms.
+		role.Rules = append(role.Rules, []rbacv1.PolicyRule{
```
do we need to add this perms to the cpo as well i.e. reconcileControlPlaneOperatorRole to honour [privilege escalation prevention](https://kubernetes.io/docs/reference/access-authn-authz/rbac/#privilege-escalation-prevention-and-bootstrapping)?

davidvossel replied:
the cpo role already has these route and service related permissions [1] [2]. I'm only transferring a subset that we need to hcco. 


1. https://github.com/openshift/hypershift/blob/bdb014332d070085d2ed8319a213c7a4c0a05f1f/hypershift-operator/controllers/hostedcluster/hostedcluster_controller.go#L2274
2. https://github.com/openshift/hypershift/blob/bdb014332d070085d2ed8319a213c7a4c0a05f1f/hypershift-operator/controllers/hostedcluster/hostedcluster_controller.go#L2289

enxebre commented on hypershift-operator/controllers/hostedcluster/hostedcluster_controller.go:
```diff
@@ -4212,6 +4218,48 @@ func (r *HostedClusterReconciler) serviceAccountSigningKeyBytes(ctx context.Cont
+func (r *HostedClusterReconciler) reconcileKubevirtPlatformDefaultSettings(ctx context.Context, hc *hyperv1.HostedCluster) error {
+	// auto generate the basedomain by retrieving the default ingress *.apps dns.
+	baseDomainPassthrough := hc.Annotations[hyperv1.KubeVirtBaseDomainPassthroughAnnotation]
+	if baseDomainPassthrough == "true" && hc.Spec.DNS.BaseDomain == "" {
```
what happens if `baseDomainPassthrough == "true" && hc.Spec.DNS.BaseDomain =!= ""`, Is that valid?

I'm +1 to make this a field and the include any invalid input combination for .baseDomainPassthrough and .BaseDomain in to `validateConfigAndClusterCapabilities`

davidvossel replied:
That scenario is not allowed during Create. After creation, the hostedcluster controller updates the BaseDomain field to the correct value based on domain name of the cluster running the VMs. 

I added a creation check `validateKubevirtBaseDomainPassthroughCreate` that ensures someone can't try to both enable the passthrough feature and set their own basedomain. That check is a part of the webhook, so it might be best effort, but i think that's probably the most feasible option for catching this configuration error.

davidvossel commented:
/hold cancel

@enxebre I updated the PR to switch back to an api field. Now, when the hostedcluster.Spec.Platform.Kubevirt.BaseDomainPassthrough option is enabled, we'll go through all the logic of setting the DNS basedomain automatically and creating the routes/services.  I added a validation check to try and catch the scenario where someone enabled this feature and tried to set their own basedomain (which isn't allowed)

davidvossel commented:
/test kubevirt-e2e-kubevirt-aws-ovn

davidvossel commented:
/test kubevirt-e2e-kubevirt-aws-ovn

orenc1 commented on api/v1beta1/hostedcluster_types.go:
```diff
@@ -626,6 +626,40 @@ type PlatformSpec struct {
+	//
+	// +optional
+	// +immutable
+	Kubevirt *KubevirtPlatformSpec `json:"kubevirt,omitempty"`
```
isn't the correct notation should be `KubeVirt`?

davidvossel replied:
for consistency I am using the same notation that we use in the NodePool api though. example...

```
        // Kubevirt specifies the configuration used when operating on KubeVirt platform.
        //
        // +optional
        Kubevirt *KubevirtNodePoolPlatform `json:"kubevirt,omitempty"`
```

I wouldn't be opposed to transitioning to use `KubeVirt` across the board within hypershift in the future. This president of using `Kubevirt` unfortunately got set within capk as part of Apple donating that code. It has carried its way into hypershift due to the KubevirtMachine objects and other apis like that. That's a larger change than i'd like to make right now though. So i'm favoring consistency instead

orenc1 commented:
I'm convinced that implementing the API field for `BaseDomainPassthrough` is a better option in this case, for more visibility and transparency to the user.
/lgtm
on my side.

orenc1 commented:
/test e2e-aws
